### ISSUE TITLE
Lifetime may not live enough regression

### ISSUE THREAD

**Issue by iamsauravsharma** [rust-lang/rust#122230]
### Code

I tried this code:

```rust
struct Test {
    field: Vec<&'static str>
}

impl Test {
    fn field<'field>(&self) -> impl Iterator<Item = &&'field str> {
        self.field.iter()
    }
}
```

I expected to see this happen: no compilation error
Instead, this happened:
```
error: lifetime may not live long enough
  --> src/lib.rs:21:9
   |
20 |     fn field<'field>(&self) -> impl Iterator<Item = &&'field str> {
   |              ------  - let's call the lifetime of this reference `'1`
   |              |
   |              lifetime `'field` defined here
21 |         self.field.iter()
   |         ^^^^^^^^^^^^^^^^^ method was supposed to return data with lifetime `'1` but it is returning data with lifetime `'field`

error: could not compile `playground` (lib) due to 1 previous error
```

It most recently works on `nightly-2024-03-06` but returns error on `nightly-2024-03-07`

**Comment by lqd** [comment]
I believe this should be due to the soundness fix in #121679.

**Comment by oli-obk** [comment]
Yes, that PR was a follow up to https://github.com/rust-lang/rust/pull/115008#issuecomment-1757775356

Your code example can be used unsoundly, even if a smarter analysis could infer the right lifetimes. 

We should have cratered the change and opened a fix PR against your repo, apologies about the unannounced breakage. I'll start a crater run now

**Comment by oli-obk** [comment]
@craterbot run name=opaque-soundness-fix start=80399064afa4a2cd153f30d02c25f7ea0383ed65 end=09bc67b9158392361780e779d32997f14cc75c39 mode=check

**Comment by oli-obk** [comment]
@craterbot run name=opaque-soundness-fix start=80399064afa4a2cd153f30d02c25f7ea0383ed65 end=09bc67b9158392361780e779d32997f14cc75c39 mode=check-only

**Comment by iamsauravsharma** [comment]
Not problem, I noticed this issue only in one of the recently released updates of third-party crates that I've been using. I've already created PR for fix, cedar-policy/cedar#712, for it.

**Comment by Mark-Simulacrum** [comment]
@craterbot abort

**Comment by Mark-Simulacrum** [comment]
@craterbot run name=opaque-soundness-fix start=master#80399064afa4a2cd153f30d02c25f7ea0383ed65 end=master#09bc67b9158392361780e779d32997f14cc75c39 mode=check-only

**Comment by oli-obk** [comment]
I opened PRs against all affected crates. The original reporter's crate did not show up since it was already fixed before crater started running
